pkg/pillar/cmd/vcomlink/vcomlink.go

```diff
@@ -0,0 +1,232 @@
+	logger = loggerArg
+	log = logArg
+
+	ctx := vcomLinkContext{
+		ps: ps,
+	}
+	agentbase.Init(&ctx, logger, log, agentName,
+		agentbase.WithPidFile(),
+		agentbase.WithBaseDir(baseDir),
+		agentbase.WithWatchdog(ps, warningTime, errorTime),
+		agentbase.WithArguments(arguments))
+
+	// Run a periodic timer so we always update StillRunning
+	stillRunning := time.NewTicker(25 * time.Second)
+	ps.StillRunning(agentName, warningTime, errorTime)
+
+	// Look for global config such as log levels
+	subGlobalConfig, err := ps.NewSubscription(pubsub.SubscriptionOptions{
+		AgentName:     "zedagent",
+		MyAgentName:   agentName,
+		TopicImpl:     types.ConfigItemValueMap{},
+		Persistent:    true,
+		Activate:      false,
+		Ctx:           &ctx,
+		CreateHandler: handleGlobalConfigCreate,
+		ModifyHandler: handleGlobalConfigModify,
+		DeleteHandler: handleGlobalConfigDelete,
+		WarningTime:   warningTime,
+		ErrorTime:     errorTime,
+	})
+	if err != nil {
+		log.Fatal(err)
+	}
+	ctx.subGlobalConfig = subGlobalConfig
+	subGlobalConfig.Activate()
+
+	for !ctx.GCInitialized {
+		log.Functionf("waiting for GCInitialized")
+		select {
+		case change := <-subGlobalConfig.MsgChan():
+			subGlobalConfig.ProcessChange(change)
+		case <-stillRunning.C:
+		}
+		ps.StillRunning(agentName, warningTime, errorTime)
+	}
+	log.Functionf("processed GlobalConfig")
+
+	// start listening on vsock
+	go startVsockServer()
+
+	for {
+		select {
+		case change := <-subGlobalConfig.MsgChan():
+			subGlobalConfig.ProcessChange(change)
+		case <-stillRunning.C:
+		}
+		ps.StillRunning(agentName, warningTime, errorTime)
+	}
+}
+
+func handleGlobalConfigCreate(ctxArg interface{}, key string,
+	statusArg interface{}) {
+	handleGlobalConfigImpl(ctxArg, key, statusArg)
+}
+
+func handleGlobalConfigModify(ctxArg interface{}, key string,
+	statusArg interface{}, oldStatusArg interface{}) {
+	handleGlobalConfigImpl(ctxArg, key, statusArg)
+}
+
+func handleGlobalConfigImpl(ctxArg interface{}, key string,
+	statusArg interface{}) {
+
+	ctx := ctxArg.(*vcomLinkContext)
+	if key != "global" {
+		log.Functionf("handleGlobalConfigImpl: ignoring %s", key)
+		return
+	}
+	log.Functionf("handleGlobalConfigImpl for %s", key)
+	gcp := agentlog.HandleGlobalConfig(log, ctx.subGlobalConfig, agentName,
+		ctx.CLIParams().DebugOverride, logger)
+	if gcp != nil {
+		ctx.GCInitialized = true
+	}
+	log.Functionf("handleGlobalConfigImpl done for %s", key)
+}
+
+func handleGlobalConfigDelete(ctxArg interface{}, key string,
+	statusArg interface{}) {
+
+	ctx := ctxArg.(*vcomLinkContext)
+	if key != "global" {
+		log.Functionf("handleGlobalConfigDelete: ignoring %s", key)
+		return
+	}
+	log.Functionf("handleGlobalConfigDelete for %s", key)
+	agentlog.HandleGlobalConfig(log, ctx.subGlobalConfig, agentName,
+		ctx.CLIParams().DebugOverride, logger)
+	log.Functionf("handleGlobalConfigDelete done for %s", key)
+}
+
+func startVsockServer() {
+	// XXX : this rudimentary vsock server, if we want to support multi-vm
```

Comment:
So in current implementation only a single vm is served?

Reply:
no, it can handle multiple VMs, I ment we can have a separate port for each service or vm to make it easier to distinguish what is what. The comment is misleading, I'll change it.